src/libpcp_web/src/timer.c

```diff
@@ -0,0 +1,222 @@
+    int				seq;
+    void			*data;
+    pmWebTimerCallBack		callback;
+    struct timerCallback	*next;
+} timerCallback;
+
+static timerCallback	*timerCallbackList;
+static uv_timer_t	pmwebapi_timer;
+static uv_mutex_t	timerCallbackMutex = PTHREAD_MUTEX_INITIALIZER;
+static struct mmv_registry *server_registry; /* global generic server metrics */
+static void		*server_map;
+
+static void
+webapi_timer_worker(uv_timer_t *arg)
+{
+    timerCallback	*timer;
+    uv_handle_t         *handle = (uv_handle_t *)arg;
+
+    (void)handle;
+    uv_mutex_lock(&timerCallbackMutex);
+    for (timer = timerCallbackList; timer; timer = timer->next) {
+    	if (timer->callback)
+	    timer->callback(timer->data);
+    }
+    uv_mutex_unlock(&timerCallbackMutex);
+}
+
+/*
+ * Register given callback function and it's private data in the
+ * global timer list. Callbacks must be non-blocking and short,
+ * e.g. to refresh instrumentation, garbage collection, etc.
+ */
+int
+pmWebTimerRegister(pmWebTimerCallBack callback, void *data)
+{
+    static int		seq = 0;
+    timerCallback	*timer = (timerCallback *)malloc(sizeof(timerCallback));
+
+    if (timer == NULL)
+    	return -ENOMEM;
+    timer->seq = seq++;
+    timer->data = data;
+    timer->callback = callback;
+    uv_mutex_lock(&timerCallbackMutex);
+    if (timerCallbackList == NULL) {
+    	uv_timer_init(uv_default_loop(), &pmwebapi_timer);
+	uv_timer_start(&pmwebapi_timer, webapi_timer_worker, 1000, 1000);
+    }
+    timer->next = timerCallbackList;
+    timerCallbackList = timer;
+    uv_mutex_unlock(&timerCallbackMutex);
+
+    return timer->seq;
+}
+
+int
+pmWebTimerRelease(int seq)
+{
+    int			sts = -EINVAL;
+    timerCallback	*timer, *prev = NULL;
+
+    uv_mutex_lock(&timerCallbackMutex);
+    for (timer = timerCallbackList; timer; prev = timer, timer = timer->next) {
+	if (timer->seq == seq) {
+	    if (timer == timerCallbackList)
+	    	timerCallbackList = timer->next;
+	    else
+	    	prev->next = timer->next;
+	    free(timer);
+	    sts = 0;
+	    break;
+	}
+    }
+    uv_mutex_unlock(&timerCallbackMutex);
+
+    return sts;
+}
+
+/* stop timer and free all timers */
+void
+pmWebTimerReleaseAll(void)
+{
+    timerCallback	*timer, *next;
+
+    uv_mutex_lock(&timerCallbackMutex);
+    uv_timer_stop(&pmwebapi_timer);
+    for (timer = timerCallbackList; timer; timer = next) {
+	next = timer->next;
+	free(timer);
+    }
+    timerCallbackList = NULL;
+    uv_mutex_unlock(&timerCallbackMutex);
+}
+
+/*
+ * timer callback to refresh generic server metrics
+ */
+static void
+server_metrics_refresh(void *map)
+{
+    double		usr, sys;
+    unsigned long	datasz = 0;
+    struct rusage	usage = {0};
+    pmAtomValue		*value;
+
+    __pmProcessDataSize(&datasz);
+    __pmProcessRunTimes(&usr, &sys);
+    usr *= 1000.0; /* milliseconds */
+    sys *= 1000.0;
+    if (getrusage(RUSAGE_SELF, &usage) < 0)
+	return;
+
+    if ((value = mmv_lookup_value_desc(map, "cpu.user", NULL)) != NULL)
```

Comment:
This code would be more efficient if it didn't lookup the MMV offset for every metric on every call.  Cache these offsets (look 'em up once for each metric during initialization/setup, then use an array indexed by 'enum server_metric' field holding pmAtomValue pointers, for direct access here) - and then you only need to call mmv_set_value during the timer.

[ Likewise throughout libpcp_web - I thought this was done awhile back?  Maybe worth opening an issue to track that I guess since it seems to have fallen through the cracks.  Or am I missing something there? ]